src/crimson/osd/osd_operations/snaptrim_event.cc

```diff
@@ -479,59 +479,64 @@ SnapTrimObjSubEvent::remove_or_update_iertr::future<>
 SnapTrimObjSubEvent::with_pg(
```

Comment:
Scanned all other places using `interruptor::with_interruption()`, they are always used at the boundary of the non-interruptive logic, in order to handle interruptions and return a non-interruptive future for non-interruptible logic.

But `SnapTrimObjSubEvent::with_pg()` here is still returning an interruptive-future (`remove_or_update_iertr::future<>`), so I'm not following why `interruptor::with_interruption()` is missing below.

Reply:
I see your point, I think maybe we should change the return type to a non-interruptible one. What do you think? Actually, this issue was that we now call seastar::yield() in `ShardServices::start_operation()`, which may leak `interrupt conditions`

Reply:
> I think maybe we should change the return type to a non-interruptible one.

Ah, that makes sense to me now.

Reply:
> Actually, the issue is that we call seastar::yield() in ShardServices::start_operation() now, which means new operations should have new and independent interrupt conditions.

Hmm, why `seastar::yield() is breaking the invariants of interruptible future, changing this pr to use interruptible yield()` was deleted?

Is it reasonable to provide a version of `start_operation_interruptive()` that wraps `seastar::yield()` with `make_interruptible()`, instead of introducing independent interrupt conditions?

Reply:
> I assume the original non-independent version was working before introducing yield(). So is it reasonable to provide a version of `start_operation_interruptive()` that wraps `seastar::yield()` with `make_interruptible()`, instead of introducing independent interrupt conditions?

Not really. To correctly preserve the interrupt condition before and after `seastar::yield()`, we need to invoke `interruptor::yield()`. However, I'm not sure whether it's appropriate to let `ShardServices` see the interruptor type and only `snap trim sub events` has this kind of issue, so I decided to modify the snap trim code. What do you think?

Reply:
> However, I'm not sure whether it's appropriate to let ShardServices see the interruptor type ... so I decided to modify the snap trim code.

I don't see why it isn't appropriate to provide an interruptible version. But I also don't see any issues with the current version.